Serve OpenAPI spec for registered CRDs

roycaihw opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
Generate and serve the OpenAPI spec for registered CRDs

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #58603

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->

reverted in https://github.com/kubernetes/kubernetes/pull/71158, original release note was:
```
kube-apiserver now serves OpenAPI specs for registered CRDs with defined validation schemata. Kubectl will validate client-side using those.
```

```release-note
NONE
```

/sig api-machinery

cc @mbohlool @sttts

roycaihw commented:
/retest

sttts commented:
Oh wow. This was long overdue. Thanks for finally fixing this.

/cc @nikhita

roycaihw commented on pkg/master/controller/crdregistration/crdregistration_controller.go:
```diff
@@ -208,19 +222,48 @@ func (c *crdRegistrationController) handleVersionUpdate(groupVersion schema.Grou
+			}
+			c.apiServiceRegistration.AddAPIServiceToSync(apiService)
+			// Serve CRD OpenAPI spec if we have non-nil OpenAPIV3Schema
+			if c.openAPIAggregationManager != nil &&
```
Yes. This serve-openapi-spec part lives in `crdRegistrationController.handleVersionUpdate()`, which gets called by `crdRegistrationController.processNextWorkItem()`. The work queue consists of all CRD adding/updating/deleting events. 

When a validation schema in a CRD changes, this part re-constructs the entire swagger (spec) for the CRD and calls `c.openAPIAggregationManager.AddUpdateLocalAPIServiceSpec()`. (Arguably it can instead read the existing spec and just update the schema, which sounds more efficient. But the construction part is just filling up a template, so I don't expect it to introduce much overhead)

roycaihw commented on pkg/master/controller/crdregistration/crdregistration_controller.go:
```diff
@@ -208,19 +222,48 @@ func (c *crdRegistrationController) handleVersionUpdate(groupVersion schema.Grou
+				crd.Spec.Validation != nil &&
+				crd.Spec.Validation.OpenAPIV3Schema != nil {
+				// Construct the OpenAPI swagger (spec) of the CRD
+				schemaV2, err := Convert_apiextensions_JSONSchemaProps_To_spec_Schema(crd.Spec.Validation.OpenAPIV3Schema)
```
Actually I'm not super confident about this. I didn't find an off-the-shelf Golang v3->v2 Schema converter ([this one](https://github.com/getkin/kin-openapi) serves v3 Schema in v2 swagger). 

I pruned `OneOf`, `AnyOf`, `AllOf` and `Not` from v3 Schema: https://github.com/kubernetes/kubernetes/pull/67205/files#diff-cb2a8286afc7142f32e0ce86b04bc5c3

sttts commented on pkg/master/controller/crdregistration/construction.go:
```diff
@@ -0,0 +1,324 @@
+limitations under the License.
+*/
+
+package crdregistration
```
this factoring is not a good idea. The master should not know about CRDs. CRDs should be normal resources, in this case provided by a delegated API server.

I could imagine to have very minimal CRD support here in the master. But all the code related to openapi certainly belong into apiextensions-apiserver, not into master.

sttts replied:
While going through the current code: we already have this minimal crd controller. But we should keep pushing logic into apiextensions-apiserver as far as possible. This swagger code certainly belongs there and not into the kube-apiserver/master.

roycaihw replied:
I agree. Moved to `apiextensions-apiserver/pkg/openapi/`

sttts commented on pkg/master/controller/crdregistration/conversion.go:
```diff
@@ -0,0 +1,55 @@
+
+// Convert_apiextensions_JSONSchemaProps_To_spec_Schema converts our internal OpenAPI v3 schema
+// (*apiextensions.JSONSchemaProps) to an OpenAPI v2 schema (*spec.Schema).
+func Convert_apiextensions_JSONSchemaProps_To_spec_Schema(in *apiextensions.JSONSchemaProps) (*spec.Schema, error) {
```
Why does this use the internal types in the first place? This is a controller and controllers should use external types (the old controllers made this horrible mistake to be based on internal types, but we are working on fixing this).

roycaihw replied:
Updated. Thanks for pointing it out. I kept a conversion in `crdregistration_controller.go` to convert the internal `CustomResourceDefinitionSpec` to `v1beta1.CustomResourceDefinitionSpec`. The rest parts use the v1beta1 types.

sttts commented on pkg/master/controller/crdregistration/crdregistration_controller.go:
```diff
@@ -208,19 +223,50 @@ func (c *crdRegistrationController) handleVersionUpdate(groupVersion schema.Grou
+			}
+			c.apiServiceRegistration.AddAPIServiceToSync(apiService)
+			// Serve CRD OpenAPI spec if we have non-nil openAPIAggregationManager
+			if c.openAPIAggregationManager != nil {
```
Can we move all of this into apiextensions-apiserver? In other words: only use the informer as a trigger that one of the delegated apiservers (which in this case happens to be apiextensions-apiserver) has updated resources and then pull the new openapi spec fromt the delegated apiserver to merge it in?

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go:
```diff
@@ -49,6 +49,16 @@ type AggregationManager interface {
+	// spec is dynamic (e.g. apiExtensionsServer), the recorded delegationAPIServices
+	// can be exposed to aggregation controller, which periodically try syncing the
+	// OpenAPI specs based on the API services' name as key
+	GetDelegationAPIServicesName() []string
```
why a slice if the func gives a name?

roycaihw replied:
Sorry it meant to give a list of names. Currently it holds the apiservice names of two delegation apiservers (generic apiserver and apiextensions apiserver). I updated the func name and documentation.

sttts commented on pkg/master/controller/crdregistration/crdregistration_controller.go:
```diff
@@ -58,17 +60,30 @@ type crdRegistrationController struct {
+	// save some code, but openapi.AggregationManager is more complex than what we need: CRD openapi
+	// spec aggregation should have no conflict so we can simply add things up, instead of walking
+	// through all swagger  to resolve conflicts. However, currently we are not optimizing performance
+	// around CRDs because we don't have a lot of CRDs in cluster. We should keep in mind that this
```
What does this mean if we have 100ish many CRDs?

roycaihw replied:
Every time one CRD gets added/updated/deleted, `openapi.AggregationManager` [builds the entire OpenAPI spec](https://github.com/kubernetes/kubernetes/blob/6d76e35b392a458c9c02e916326bba8e9c8ab8f4/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator.go#L188-L215) by 1. sorting the slice of spec by priority (not needed for CRDs because CRD apiservices are of same priority); 2. iterating through the specs and merging each spec (`source`) with the previous merged specs (`dest`). 

In each [merge process](https://github.com/kubernetes/kube-openapi/blob/e3762e86a74c878ffed47484592986685639c2cd/pkg/aggregator/aggregator.go#L256-L374) the aggregator iterates through `source.Paths` and `source.Definitions` a couple of times to check if there are any conflicts against `dest`, but it's also unnecessary for CRDs because CRD GVKs are unique. 

What we need can be something simple like: 
```
for k, v := range source.Definitions {
	dest.Definitions[k] = v
}
for k, v := range source.Paths.Paths {
	dest.Paths.Paths[k] = v
}
```

The main benefit of reusing `openapi.AggregationManager` is to save some code on recording the apiservices and specs. I will add another AggregationManager for CRDs

sttts replied:
"total 0.76s" sounds far too much. In any case, we should probably throttle this to once every few seconds. Wdyt?

roycaihw replied:
0.76s is the aggregation time for in-tree specs, which only happens at kube-apiserver start time. 

But I agree that we should have the simplified aggregator that doesn't check conflict. If we have 100ish CRDs and they get added to the cluster at start time, we will have a peak hitting the aggregation way more frequent than once every few seconds.

sttts commented on pkg/master/controller/crdregistration/conversion.go:
```diff
@@ -0,0 +1,55 @@
+		return nil, err
+	}
+	// Remove unsupported fields in OpenAPI v2
+	out.AllOf = nil
```
@nikhita had a nice spreadsheet with the differences.

roycaihw replied:
@nikhita Could you point me to it? Thanks!

sttts commented:
We had the following discussion before. But now with more and more complexity I want to bring it up again:

Why do we need special CRD logic in the kube-apiserver to merge OpenAPI specs? Why are CRDs different from aggregated API servers?

The only answer I see is that CRDs might change more frequently than aggregated API server. So having a trigger logic using a CRD informer is probably fine. But the actual logic around openapi spec retrieval should not be different. But it is in the current code base and even more so with this PR. I don't think we should go this route, but rethink how to unify it with other aggregated API servers.

roycaihw commented:
> Why do we need special CRD logic in the kube-apiserver to merge OpenAPI specs? Why are CRDs different from aggregated API servers?

@sttts Also try to reply https://github.com/kubernetes/kubernetes/pull/67205#discussion_r211182550 here. 

You are right. CRDs (apiextensions apiserver) aren't different from aggregated API servers. The trigger logic in crdregistration_controller only triggers an `openapi.AggregationManager` that updates the OpenAPI spec served on apiextensions apiserver. 

The problem is currently aggregator apiserver treats the OpenAPI specs from delegation targets as static and never re-pulls them. In this PR I added delegation target API services to queue, so that aggregator apiserver treats them the same as other aggregated API servers and periodically pulls the OpenAPI spec. https://github.com/kubernetes/kubernetes/pull/67205/files#diff-496d390e5ec105be6afa681750c80e70R88

sttts commented:
> In this PR I added delegation target API services to queue, so that aggregator apiserver treats them the same as other aggregated API servers and periodically pulls the OpenAPI spec.

This sounds fine. I am also fine with a CRD informer to trigger an update in realtime. But I would prefer not to see any CRD logic there which exceeds the pure trigger. The actual openapi schema logic for CRDs should be contained in apiextensions-apiserver and not leak outside.

roycaihw commented:
> I am also fine with a CRD informer to trigger an update in realtime. 

Yes that's an alternative. Currently in this PR aggregator apiserver pulls openapi specs from delegation targets [every minute](https://github.com/kubernetes/kubernetes/blob/ddebb2480ddcd976071a337c8915c4583f31ae41/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go#L130-L139) on my test cluster.

> But I would prefer not to see any CRD logic there which exceeds the pure trigger. The actual openapi schema logic for CRDs should be contained in apiextensions-apiserver and not leak outside.

I moved the openapi logic to apiextensions-apiserver. The CRD informer now makes one function call to get the openapi spec and etag, then calls `c.openAPIAggregationManager.AddUpdateLocalAPIServiceSpec` to update the spec served by apiextensions apiserver.

sttts commented on cmd/kube-apiserver/app/aggregator.go:
```diff
@@ -132,7 +133,8 @@ func createAggregatorServer(aggregatorConfig *aggregatorapiserver.Config, delega
 		apiExtensionInformers.Apiextensions().InternalVersion().CustomResourceDefinitions(),
-		autoRegistrationController)
+		autoRegistrationController,
+		crdOpenAPIAggregationManager)
```
is this only used for CRDs?

roycaihw replied:
Yes. `crdOpenAPIAggregationManager` holds the openapi services of apiextensions apiserver. https://github.com/kubernetes/kubernetes/pull/67205/files#diff-36e1d7088209b1ab6ff878b18bb35210R200

sttts commented on pkg/master/controller/crdregistration/crdregistration_controller.go:
```diff
@@ -208,19 +224,44 @@ func (c *crdRegistrationController) handleVersionUpdate(groupVersion schema.Grou
+				if err := v1beta1.Convert_apiextensions_CustomResourceDefinitionSpec_To_v1beta1_CustomResourceDefinitionSpec(&crd.Spec, versionedSpec, nil); err != nil {
+					return err
+				}
+				crdspec, etag, err := apiextensionsopenapi.CustomResourceDefinitionOpenAPISpec(versionedSpec)
```
This is "the one call into apiextensions-apiserver"?

Wondering, could we push the whole openapi spec generation for all CRDs into apiextensions-apiserver and then only merge it into the kube-apiserver openapi spec?

Think of apiextensions-apiserver running standalone. The same code path should be used in that case.

roycaihw replied:
I think there are three parts of the problem:

1. Generating the spec for a new added/updated CRD: 

This can happen at any place along the CRD adding/updating path. Currently in this PR we put it at where the KAS discovery registration happens, but we should be able to move it to `k8s.io/apiextensions-apiserver` (maybe in one of the [CRD controllers](https://github.com/kubernetes/kubernetes/blob/c0be4cc09c5a685331cc310273cfdd633eac54ac/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go#L196-L199)).

2. Serving (aggregating) the spec to apiextensions apiserver:

This requires writing to the OpenAPIService that gets hooked to `apiExtensionsServer.GenericAPIServer.Handler`. Currently the [OpenAPIService](https://github.com/kubernetes/kubernetes/blob/c0be4cc09c5a685331cc310273cfdd633eac54ac/cmd/kube-apiserver/app/server.go#L193-L194) gets created after [apiextensions apiserver](https://github.com/kubernetes/kubernetes/blob/c0be4cc09c5a685331cc310273cfdd633eac54ac/cmd/kube-apiserver/app/server.go#L179) gets created. 

To use the same code path for standalone apiextensions apiserver, code in `k8s.io/apiextensions-apiserver` needs to have access to the OpenAPIService. I'm thinking how can we pass it to one of the CRD controllers. @sttts Any suggestions?

3. Aggregating the apiextensions apiserver spec to aggregator apiserver spec:

This should happen in staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go

roycaihw commented:
@sttts I moved the logic to apiextensions-apiserver and implemented another openapi spec aggregator that does not check path/definition conflicts for CRD. PTAL

roycaihw commented:
/milestone v1.12

mbohlool commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator.go:
```diff
@@ -188,6 +203,7 @@ func sortByPriority(specs []openAPISpecInfo) {
 // buildOpenAPISpec aggregates all OpenAPI specs.  It is not thread-safe. The caller is responsible to hold proper locks.
 func (s *specAggregator) buildOpenAPISpec() (specToReturn *spec.Swagger, err error) {
 	specs := []openAPISpecInfo{}
+
```
remove whitespace

mbohlool commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go:
```diff
@@ -72,6 +78,12 @@ func NewAggregationController(downloader *Downloader, openAPIAggregationManager
 	}
 
 	c.syncHandler = c.sync
+	if services := openAPIAggregationManager.GetDelegationAPIServiceNames(); services != nil {
```
In AddAPIService, there is an if statement that skips APIService if the service is nil. I assume those are local API services aka DelegateAPIServices, right? Can't you just remove that if statement?

roycaihw replied:
The delegation targets' APIServices/OpenAPISpec aggregation happens in [BuildAndRegisterAggregator](https://github.com/kubernetes/kubernetes/blob/cd06419973297dd1f65d1e47e9a3363bfbbcc899/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator.go#L72-L73) which happens before `AggregationController` is created. 

I think `func (c *AggregationController) AddAPIService` only gets called on new/updated APIService which triggers apiServiceInformer, but it's not called on delegation targets.

mbohlool commented on staging/src/k8s.io/apiextensions-apiserver/pkg/openapi/BUILD:
```diff
@@ -0,0 +1,46 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
```
Move to generated

mbohlool commented on staging/src/k8s.io/apiextensions-apiserver/pkg/openapi/construction.go:
```diff
@@ -0,0 +1,327 @@
+				// TODO(roycaihw): this is a hack to let apiExtension apiserver and generic kube-apiserver
+				// to have the same io.k8s.api.autoscaling.v1.Scale definition, so that aggregator server won't
+				// detect name conflict and create a duplicate io.k8s.api.autoscaling.v1.Scale_V2 schema
+				// when aggregating the openapi spec. This hack won't tolerate any changes in
```
what do you mean by won't tolerate? does it mean failure or copying to _V2?

CaoShuFeng commented:
Hi, I can't understand the output of this line:

https://gist.github.com/roycaihw/1bb6eb7973b5496cb501ee50c8ee1ee3#file-crd_openapi_spec-diff-L799

Why FooList.items is referenced to io.k8s.api.autoscaling.v1.Scale?

CaoShuFeng commented on staging/src/k8s.io/apiextensions-apiserver/pkg/openapi/construction.go:
```diff
@@ -0,0 +1,327 @@
+		RespondsWith(200, okResponse(ref)).
+		RespondsWith(201, createdResponse(ref)).
+		RespondsWith(202, acceptedResponse(ref)).
+		AddParam((&spec.Parameter{ParamProps: spec.ParamProps{Schema: spec.RefSchema(deleteOptionsSchemaRef)}}).
```
Why would a create request have a deleteOptions in body?

roycaihw replied:
I must be copy-pasting code around. Fixed schema ref in create, replace and patch operations.

roycaihw commented:
> Why FooList.items is referenced to io.k8s.api.autoscaling.v1.Scale?

@CaoShuFeng Nice catch! I used wrong [ResourceKind](https://github.com/kubernetes/kubernetes/pull/67205/commits/c160bbd45cc2f3ca6d16643580f2b1fca115abce#diff-13e90024c6d4134dc6478d2b424b2511R308) when constructing schema ref. It should be `Scale`. Updated

dims commented:
/test pull-kubernetes-kubemark-e2e-gce-big

tpepper commented:
SIG leads, can we "/milestone clear" this?  It's much too late in the cycle now as we're truly working to only merge a few last critical bug fixes.

roycaihw commented:
@tpepper Yes. This is a bug fix that doesn't belong to any feature. We can cherry-pick it to 1.12 if we want it later. I don't have the permission to change the milestone though. @mbohlool @sttts could you help make the change please? Thanks!

tpepper commented:
/milestone clear

roycaihw commented:
@ecordell Thanks for reminding! I lost track of this after 1.12 was cut. Let me take a look to see if there are any comments left un-addressed and then request for another round of review.

sttts commented:
Would like to see this in 1.13 finally.

sttts commented:
/cc @deads2k @liggitt @mbohlool

smarterclayton commented:
This is covered under existing stabilization efforts around CRDs and is an important bug/gap.

/approve

AishSundar commented:
/milestone v1.13

/cc @nikopen

roycaihw reviewed (commented):
Sorry for the delay of updating the status. I took a look and found the following two places that I want to improve on (previous comments should have been addressed already-- please let me know if I missed any).

In the meanwhile, please take a look and leave any comments/suggestions (I rearranged the commits to make them easier for review).

/cc @deads2k @liggitt @mbohlool @sttts

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/openapi/conversion.go:
```diff
@@ -0,0 +1,48 @@
+	out.AllOf = nil
+	out.OneOf = nil
+	out.AnyOf = nil
+	out.Not = nil
```
During pruning we are allowed to make the specification weaker, not stricter. This means:

- we can drop clauses if they are positive (in the sense of `A`, `not not A`, `not not not not A`: `A` is positive in each of them. In contrast: `not A`, `not not not A`: here `A` is negative.
- pruning in a negative position means replacing a clause by False (= `{not: {}}` in OpenAPI).
- dropping all 4 is fine.
- both `not` and `oneOf` (equals `(andOf (anyOf A B) (not (andOf A B)))`) make pruning complex as you have to compute for each to-be-dropped clause in a subformular, whether it is positive or negative.

Hence, the easiest solution is to drop all 4.

We could do a bit better with `AllOf` by merging sub-formulars into one schema, e.g. `{allOf {type: number} {minValue: 42.0}}` can be merged into `{type: number, minValue: 42.0}`.

roycaihw replied:
I realized that `allOf` was supported by OpenAPI Spec v2: https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#schema-object (we didn't use `allOf` for k8s built-in APIs). `anyOf`, `oneOf` and `not` are new to v3. We can probably keep `allOf` and drop the other three. 

Sorry I didn't understand why we want to merge sub-formulas of `allOf` into one schema. Could you explain more on that?

sttts replied:
I assumed `allOf` is not part of v2. If it is, forget my comment about merging. Not needed.

roycaihw commented:
/retest

roycaihw commented:
@sttts @mbohlool This PR is ready for another round of review. It now 
- supports CRD with per-version schema
- keeps `allOf` during v3->v2 conversion

I tested locally using a CRD with per-version schema (attached below). [Here is a gist](https://gist.github.com/roycaihw/8ae7519a2de061c7b33c5cd4d1a1e2df) containing changes in openapi spec after the CRD was created. The openapi spec was restored after the CRD got deleted. 
```
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: foos.samplecontroller.k8s.io
spec:
  group: samplecontroller.k8s.io
  version: v1alpha1
  versions:
    - name: v1alpha1
      storage: true
      served: true
      schema:
        openAPIV3Schema:
          description: Alpha
    - name: v1beta1
      storage: false
      served: true
      schema:
        openAPIV3Schema:
          description: Beta
  names:
    kind: Foo
    plural: foos
  scope: Namespaced
```

mbohlool commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery_controller.go:
```diff
@@ -119,6 +126,31 @@ func (c *DiscoveryController) sync(version schema.GroupVersion) error {
+					return err
+				}
+			}
+			crdspec, etag, err := apiextensionsopenapi.CustomResourceDefinitionOpenAPISpec(&crd.Spec, version.Version, versionedSchema)
```
Add a comment here that explains we do not to continue even if versionedSchema is nil as it maybe a removal of the schema for this CRD.

mbohlool commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery_controller.go:
```diff
@@ -180,6 +218,12 @@ func (c *DiscoveryController) sync(version schema.GroupVersion) error {
+		if c.openAPIAggregationManager != nil {
+			// Remove the local API service for the CRD in apiExtensionsServer's
+			// openAPIAggregationManager. If the API service doesn't exist (e.g. because
+			// CRD Validation OpenAPIV3Schema is nil), no error will be returned.
```
is the intension is to not returning error or you just saying the RemoveAPISerivceSpec does not return error. I don't think it is in the contract of that function to never return error, it is just current implementation. If we need not to return error here, we should do so by ignoring returned error from that function otherwise revise the comment please.

mbohlool commented on staging/src/k8s.io/apiextensions-apiserver/pkg/openapi/aggregator.go:
```diff
@@ -0,0 +1,227 @@
+func mergeSpecs(dest, source *spec.Swagger) {
+	// Paths may be empty, due to [ACL constraints](http://goo.gl/8us55a#securityFiltering).
+	if source.Paths == nil {
+		// If Path is nil, none of the model defined in Definitions is used and we
```
This should not happened for CRDs, right? use the proper logutil for things that should never happened.

mbohlool reviewed (commented):
Mostly nit comments.

roycaihw commented:
Comments from @mbohlool addressed. I will add an e2e test in a followup PR. 

Talked with @mbohlool offline, we decided to add a check of the feature gate `apiextensionsfeatures.CustomResourceValidation` in `customresource_discovery_controller.go`. 

Given that we had [a bug](https://github.com/kubernetes/kube-openapi/issues/66) in openapi aggregation which could keep apiserver panic during bootstrap, it's safer to have a way to disable the CRD openapi spec aggregation. 

@sttts Any other comments?

roycaihw commented:
/test pull-kubernetes-verify

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator.go:
```diff
@@ -67,6 +73,15 @@ func (s *specAggregator) addLocalSpec(spec *spec.Swagger, localHandler http.Hand
 	}
+	// Record delegation target API services' names
+	if localHandler != nil {
+		s.delegationAPIServices = append(s.delegationAPIServices, name)
```
can't this be derived from `s.openAPISpecs[name]`? Why an extra slice?

roycaihw replied:
We want to pass a list of names to the controller's [queue](https://github.com/kubernetes/kubernetes/blob/62060e1f046a888c475f656193219401fae35a0e/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go#L82), so that it can periodically sync openapi specs from delegation apiservers. ~~We didn't have a place that recorded the names of the delegation APIServices.~~

Actually, I thought we cannot use the keys (names) in `s.openAPISpecs`, because they consist of both delegation APIServices' names and aggregated APIServices' names, and there will be race condition if we pass all the keys to the queue. It turns out that we do the passing during controller initialization, so there should be no aggregated APIService's name.

I agree that the keys in `s.openAPISpecs` will be mostly the same as `s.delegationAPIServices ` during [controller initialization](https://github.com/kubernetes/kubernetes/blob/62060e1f046a888c475f656193219401fae35a0e/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go#L82) (except the [local spec with no handler](https://github.com/kubernetes/kubernetes/blob/f48e18faa4dc035cc927c6a2b34c83c8475b55fa/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator.go#L87-L88)-- it's a minor case, the controller will [drop the name](https://github.com/kubernetes/kubernetes/blob/f48e18faa4dc035cc927c6a2b34c83c8475b55fa/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go#L121-L123))

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator.go:
```diff
@@ -53,6 +53,12 @@ type specAggregator struct {
+	// spec is dynamic (e.g. apiExtensionsServer), the recorded delegationAPIServices
+	// can be exposed to aggregation controller, which periodically try syncing the
+	// OpenAPI specs based on the API services' name as key
+	delegationAPIServices []string
```
We have two names for the same thing: local and delegation api service, don't we?

roycaihw replied:
Right. "Local" APIServices contain delegation APIServices plus an extra [static spec with no handler](https://github.com/kubernetes/kubernetes/blob/f48e18faa4dc035cc927c6a2b34c83c8475b55fa/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator.go#L87-L88) IMO. 

I think we should be able to drop the `delegationAPIServices` slice as you pointed out in the other comment https://github.com/kubernetes/kubernetes/pull/67205#discussion_r233735114

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go:
```diff
@@ -72,6 +78,10 @@ func NewAggregationController(downloader *Downloader, openAPIAggregationManager
 	c.syncHandler = c.sync
+	// Add delegation target API services to queue, to aggregate dynamic OpenAPI spec from delegation target
+	for _, name := range openAPIAggregationManager.GetDelegationAPIServiceNames() {
+		c.queue.AddAfter(name, time.Second)
```
why after a second? Why not immediately?

roycaihw replied:
Was following the [existing code](https://github.com/kubernetes/kubernetes/blob/f48e18faa4dc035cc927c6a2b34c83c8475b55fa/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go#L156) for aggregating an APIService. 

I don't see strong benefit of doing one way verses the other when we add/update an APIService. The delaying interface makes more sense when we deal with [sync reaction](https://github.com/kubernetes/kubernetes/blob/f48e18faa4dc035cc927c6a2b34c83c8475b55fa/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go#L114-L123) to avoid hot-loop. 

Maybe we chose to wait a second because we have just aggregated the spec and don't want to pull the HTTP handler again immediately (we will likely get the same ETag).

sttts replied:
Just want to make sure this second is not a magic number, i.e. it breaks with zero or if the bootstrapping takes longer than the given time.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go:
```diff
@@ -191,7 +192,18 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
 
-	crdController := NewDiscoveryController(s.Informers.Apiextensions().InternalVersion().CustomResourceDefinitions(), versionDiscoveryHandler, groupDiscoveryHandler)
+	// This will wire up openapi for extension api server
+	s.GenericAPIServer.PrepareRun()
```
this does not belong here. It is to be called shortly before run as a last chance before run to wire things.

roycaihw replied:
Good point. I put it here because I need the openapi route installation ready, before passing the openapi services and spec to a controller that watches CRD and operates on the openapi services w/ spec.

Maybe we can do either of:
1. separate [openapi route installation](https://github.com/kubernetes/kubernetes/blob/f48e18faa4dc035cc927c6a2b34c83c8475b55fa/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go#L237-L241) out from `PrepareRun`, so we can wire up openapi earlier but keep everything else as it used to be: I'm not sure if it's safe/logically-correct to do so
2. use/create another controller that watches CRD after things are [wired up](https://github.com/kubernetes/kubernetes/blob/f48e18faa4dc035cc927c6a2b34c83c8475b55fa/cmd/kube-apiserver/app/server.go#L180-L185) for extension api server: I think it's similar to one of my previous approach. It has the problem of having more CRD-related code out of k8s.io/apiextension-apiserver package.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -119,6 +121,12 @@ type GenericAPIServer struct {
 
+	// Expose the registered OpenAPI Services and built static OpenAPI spec
+	// if openAPIConfig is non-nil
+	OpenAPIService          *handler.OpenAPIService
```
add comments that the former is about /swagger.json and the latter is /openapi/v2

sttts replied:
done

sttts commented on staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go:
```diff
@@ -30,17 +32,27 @@ type OpenAPI struct {
+	// with this static spec on the registered OpenAPI services
+	spec, err = builder.BuildOpenAPISpec(c.RegisteredWebServices(), oa.Config)
+	if err != nil {
+		klog.Fatalf("Failed to build open api spec for root: %v", err)
```
we should add an error return value instead

sttts replied:
Ok, we have some more already below. Nothing new.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/openapi/construction.go:
```diff
@@ -0,0 +1,324 @@
+}
+
+// ConstructCRDOpenAPISpec constructs the complete OpenAPI swagger (spec).
+func (c *SwaggerConstructor) ConstructCRDOpenAPISpec() *spec.Swagger {
```
am a bit surprised this does not share anything with the native resource or kube-openapi.

roycaihw replied:
kube-openapi requires constructing a `*restful.WebService` to build the spec. We have a huge [function](https://github.com/kubernetes/kubernetes/blob/b1a52a38e9e3651680655416cc7afbec5e119854/staging/src/k8s.io/apiserver/pkg/endpoints/installer.go#L170-L874) doing this for native resources. Now I look back and think maybe we can refactor it and reuse some code. I will give it a try as a follow-up.

Efficiency of filling a template might be another minor reason

sttts replied:
done

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/openapi/construction.go:
```diff
@@ -0,0 +1,324 @@
+							Parameters: pathParameters(),
+						},
+					},
+					fmt.Sprintf("%s/{name}/status", basePath): {
```
why doesn't this depend on the subresource property in the CRD spec?

sttts commented on hack/.golint_failures:
```diff
@@ -472,6 +472,7 @@ staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server
 staging/src/k8s.io/apiextensions-apiserver/pkg/controller/finalizer
 staging/src/k8s.io/apiextensions-apiserver/pkg/controller/status
 staging/src/k8s.io/apiextensions-apiserver/pkg/features
+staging/src/k8s.io/apiextensions-apiserver/pkg/openapi
```
why not fix it right away?

roycaihw replied:
golint was complaining about how we name the conversion functions with underscore: `Convert_v1beta1_JSONSchemaProps_To_spec_Schema` https://github.com/kubernetes/kubernetes/pull/67205/commits/6dca23fc44e4f9d9fb3bcb0227126db6f46e27eb#diff-19681f6fc979e9acc3ee091c79e3a180R30

roycaihw committed:
Enable aggregator apiserver resyncing openapi spec

from delegation apiservers

roycaihw committed:
Allow apiextensions-apiserver to import kube-openapi

roycaihw committed:
Apiextensions-apiserver aggregates CRD schemas

efficiently without checking conflicts, and wire up CRD discovery
controller to serve OpenAPI spec.

roycaihw committed:
Convert and construct OpenAPI v2 spec from CRD

validation OpenAPI v3 Schema

roycaihw committed:
generated

sttts commented:
Lgtm when green.

sttts commented:
Needs an e2e test.

roycaihw committed:
add basic e2e test and integration for CRD openapi

roycaihw commented:
Tests are green now. Pushing another commit with basic integration test and e2e test to see how they go

sttts commented:
/lgtm

smarterclayton commented:
I wasn't able to see this from a casual scan of the code, but do we have a way to pass description for fields and model objects back from this?  If not, we definitely need that in order to make `kubectl explain` be able to help users (which is one of the key consumers of openapi).

sttts commented:
@smarterclayton the description of a scheme prop is copied into the published schema. So from the code it looks like this should work.

@roycaihw can you add an e2e test which calls kubectl explain against such a CRD?

sttts commented:
@smarterclayton this shows up in the openapi spec (but kubectl explain does not do anything, should it?)

```
   "samplecontroller.k8s.io.v1alpha1.Foo": {
    "properties": {
     "spec": {
      "properties": {
       "replicas": {
        "description": "replicas description",
        "type": "integer",
        "maximum": 10,
        "minimum": 1,
        "externalDocs": {
         "description": "external replicas description",
         "url": "http://k8s.io"
        },
        "example": 42
       }
      }
     }
    }
   },
```

liggitt commented:
looks like the individual item does not include a `x-kubernetes-group-version-kind` stanza

liggitt commented:
interestingly, it looks like the `<CustomKind>List` object schema does include the gvk

smarterclayton commented:
The end use case is that a CRD is able to show something like the following
for any CRD property:

explain pods.spec.serviceAccountName
KIND:     Pod
VERSION:  v1

FIELD:    serviceAccountName <string>

DESCRIPTION:
     ServiceAccountName is the name of the ServiceAccount to use to run this
     pod. More info:

https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/

liggitt commented:
this adds the GVK to the openapi spec:

```diff
diff --git a/staging/src/k8s.io/apiextensions-apiserver/pkg/openapi/construction.go b/staging/src/k8s.io/apiextensions-apiserver/pkg/openapi/construction.go
index 804681da1f..0327e6eacb 100644
--- a/staging/src/k8s.io/apiextensions-apiserver/pkg/openapi/construction.go
+++ b/staging/src/k8s.io/apiextensions-apiserver/pkg/openapi/construction.go
@@ -96,6 +96,11 @@ func (c *SwaggerConstructor) ConstructCRDOpenAPISpec() *spec.Swagger {
        var schema spec.Schema
        if c.schema != nil {
                schema = *c.schema
+               schema.AddExtension("x-kubernetes-group-version-kind", map[string]string{
+                       "group":   c.group,
+                       "kind":    c.kind,
+                       "version": c.version,
+               })
        }
```

but explain is still confused

roycaihw commented:
@liggitt The format of `x-kubernetes-group-version-kind` in my code was wrong. It should be a list of map, instead of a single map. Fixing

roycaihw commented:
kubectl explain and validation works now. Will send a PR with an e2e test

```
$ kubectl explain csinodeinfos
KIND:     CSINodeInfo
VERSION:  csi.storage.k8s.io/v1alpha1

DESCRIPTION:

FIELDS:
   spec	<Object>
     Specification of CSINodeInfo

   status	<Object>
     Status of CSINodeInfo

```

```
$ echo '{"kind":"CSINodeInfo","apiVersion":"csi.storage.k8s.io/v1alpha1","metadata":{"name":"foo"},"spec":{"foo":true}}' \
  | kubectl create -f -
error: error validating "STDIN": error validating data: [ValidationError(CSINodeInfo): unknown field "apiVersion" in csi.storage.k8s.io.v1alpha1.CSINodeInfo, ValidationError(CSINodeInfo): unknown field "kind" in csi.storage.k8s.io.v1alpha1.CSINodeInfo, ValidationError(CSINodeInfo): unknown field "metadata" in csi.storage.k8s.io.v1alpha1.CSINodeInfo, ValidationError(CSINodeInfo.spec): unknown field "foo" in csi.storage.k8s.io.v1alpha1.CSINodeInfo.spec]; if you choose to ignore these errors, turn validation off with --validate=false
```
